### ISSUE TITLE
[Question] Path Prefixing

### ISSUE THREAD

**Issue by iMusicJJ** [PlatformCommunity/flysystem-bunnycdn#53]
Hello!

How do  you set-up the path prefixing?

I've attempted to set it up with Laravel, and it uploads the files as expected in the prefixed path, but upon URL retrieval it throws:
"This driver does not support retrieving URLs.",

Do you have an example with a prefixed path?

**Comment by sifex** [comment]
Hey, sure thing, 

Should be as simple as wrapping your `BunnyCDNAdapter` inside of a `PathPrefixAdapter`

So update

```php
            $adapter = new BunnyCDNAdapter(
                new BunnyCDNClient(
                    $config['storage_zone'],
                    $config['api_key'],
                    $config['region']
                ),
                'http://testing.b-cdn.net' # Optional
            );
```

to be

```php
            $adapter = new PathPrefixedAdapter(
                new BunnyCDNAdapter(
                    new BunnyCDNClient(
                        $config['storage_zone'],
                        $config['api_key'],
                        $config['region']
                    ),
                    'http://testing.b-cdn.net' # Optional
                ),
                'prefix'
            );
```

(Disclaimer) I wrote it by hand so no idea if it works. It's got test coverage though

**Comment by iMusicJJ** [comment]
Funky, this is what I have I believe?

```php
$adapter = new BunnyCDNAdapter(
    new BunnyCDNClient(
        $config['storage_zone'],
        $config['api_key'],
        $config['region']
    ),
    $config['pullzone_url'] ?? null
);

$adapter = new PathPrefixedAdapter($adapter, 'images/mails');
```

Which is what throws the exception when trying to generate the URL.

Full example:

```php
Storage::extend('bunnycdn-mail', function (Application $app, array $config) {
            $adapter = new BunnyCDNAdapter(
                new BunnyCDNClient(
                    $config['storage_zone'],
                    $config['api_key'],
                    $config['region']
                ),
                $config['pullzone_url'] ?? null
            );

            $adapter = new PathPrefixedAdapter($adapter, 'images/mails');

            return new FilesystemAdapter(
                new Filesystem($adapter, $config),
                $adapter,
                $config);
        });
```

**Comment by sifex** [comment]
Let me have a quick look now for ya

**Comment by sifex** [comment]
So you're not going to be able to get nice IDE completion, but there is a fall-through for calling adapter methods on the `Storage` facade. 

<img width="848" alt="Screenshot 2022-11-24 at 5 46 39 pm" src="https://user-images.githubusercontent.com/2058557/203842810-1f607edd-384e-4689-b699-9be02537b177.png">

tldr; You can call `Storage::publicUrl` instead of `Storage::url` and that seems to work for the moment.

**Comment by iMusicJJ** [comment]
Oohhh, this is where it get's funky.
I don't have access to the `Storage::url` call, it's happening inside [mailcoach](https://mailcoach.app)

I am trying to store email images on Bunny, instead of locally.

**Comment by sifex** [comment]
Huh.... Well.

Because `PathPrefixedAdapter` doesn't have a `getUrl` method (I think `v3` changed the method name), it's gonna throw "This driver does not support retrieving URLs.".


<img width="871" alt="Screenshot 2022-11-24 at 6 01 18 pm" src="https://user-images.githubusercontent.com/2058557/203845162-292507bc-1937-45eb-abc5-b995fbd37e24.png">


**The quick & dirty:**

Best way I'm going to suggest is to extend the `PathPrefixAdapter` with a getUrl method.

```php
# /app/Providers/Support/PathPrefixedAdapter.php

<?php

namespace App\Providers\Support;

use League\Flysystem\Config;

class PathPrefixedAdapter extends \League\Flysystem\PathPrefixing\PathPrefixedAdapter
{
    public function getUrl($path): string
    {
        return $this->publicUrl($path, new Config());
    }
}

```

Then use that `PathPrefixedAdapter` instead of the one provided by `\League\Flysystem`


**The actual fix:** 

`src/Illuminate/Filesystem/FilesystemAdapter.php` needs a lot of attention given the advances in Flysystem v3

**Comment by sifex** [comment]
I'll open a quick PR for Laravel now

**Comment by iMusicJJ** [comment]
Ah great! Thanks for the amazingly quick response!

**Comment by sifex** [comment]
This is expected with `PathPrefixedAdapter`: Laravel's `url()`/`temporaryUrl()` don't use Flysystem's `PublicUrlGenerator`/`TemporaryUrlGenerator` interfaces — they check `method_exists($adapter, 'getUrl'/'getTemporaryUrl')`, which the wrapper doesn't implement, hence "This driver does not support retrieving URLs."

Since **v3.4.0** the adapter has native root support that scopes *all* operations including URL generation, so drop the wrapper entirely and just set the `root` config:

```php
'bunnycdn' => [
    'driver' => 'bunnycdn',
    // ...
    'root' => env('BUNNYCDN_ROOT', 'assets'), // optional root path
],
```

Standalone: `new BunnyCDNAdapter($client, 'https://testing.b-cdn.net', 'assets')` — see the "Usage with a Root Path" section of the README.
